framework/src/constraints/ADMortarConstraint.C

```diff
@@ -170,3 +170,69 @@ ADMortarConstraint::computeJacobian(Moose::MortarType mortar_type)
 
   _assembly.processDerivatives(residuals, dof_indices, _matrix_tags, local_functor);
 }
+
+#ifdef MOOSE_GLOBAL_AD_INDEXING
+void
+ADMortarConstraint::trimDerivative(const dof_id_type & remove_derivative_index, ADReal & var)
+{
+  auto md_it = var.derivatives().nude_data().begin();
+  auto mi_it = var.derivatives().nude_indices().begin();
+
+  auto d_it = var.derivatives().nude_data().begin();
+
+  for (auto i_it = var.derivatives().nude_indices().begin();
+       i_it != var.derivatives().nude_indices().end();
+       ++i_it, ++d_it)
+    if (*i_it != remove_derivative_index)
+    {
+      *mi_it = *i_it;
+      *md_it = *d_it;
+      ++mi_it;
+      ++md_it;
+    }
+
+  std::size_t n_indices = md_it - var.derivatives().nude_data().begin();
+  var.derivatives().nude_indices().resize(n_indices);
+  var.derivatives().nude_data().resize(n_indices);
+}
+
+void
+ADMortarConstraint::trimInteriorNodeDerivatives(
+    const std::map<unsigned int, unsigned int> & domain_ip_lowerd_map,
+    std::array<MooseVariable *, 3> & moose_var,
+    ADReal & var1,
+    ADReal & var2,
+    ADReal & var3,
+    const bool is_secondary)
+{
+  // Remove interior node variable's derivatives from AD objects.
```

Comment:
Can we assert that the moose variables are nodal somewhere in this method?

Reply:
Should be done.